Add a function to get a Speed Vector from a GPS.

llessieux opened the pull request:
Add new wb_gps_get_speed_vector function to the GPS to retrieve the motion delta between the sampling points.

This pull-request fixes issue [#3660](https://github.com/cyberbotics/webots/issues/3660)

  - Add wb_gps_get_speed_vector and wired it to WbGPS and other parts.
  - Update documentation
  - Update gps device sample code to show how to get the new values.

**Documentation**
This PR updates the GPS page https://cyberbotics.com/doc/reference/gps?version=develop#gps

I hope that I got all the parts that needs to be wired.
I wasn't sure about the Matlab part. I left the function commented like the get_value ones. Not sure why it is that way in the first place but I figured that any function returning a pointer being commented the new one should be too.

Let me know if I forgot anything, quite probable.

omichel commented on docs/reference/changelog-r2021.md:
```diff
@@ -9,6 +9,7 @@ Released on XX, XXth, 2021.
     - Improved performance on [Lidar](lidar.md) point cloud generation ([#3499](https://github.com/cyberbotics/webots/pull/3499)).
     - Improved the user experience when using the object rotation around axis feature ([#3540](https://github.com/cyberbotics/webots/pull/3540)).
     - Increased the mouse wheel speed when zooming the 3D window ([#3565](https://github.com/cyberbotics/webots/pull/3565)).
+    - Add Speed Vector output to GPS ([#3742](https://github.com/cyberbotics/webots/pull/3742))
```
```suggestion
    - Added speed vector output to GPS ([#3742](https://github.com/cyberbotics/webots/pull/3742))
```

omichel commented on docs/reference/gps.md:
```diff
@@ -165,6 +168,8 @@ The `gpsReference` field of the [WorldInfo](worldinfo.md) node can be used to de
 
 The `wb_gps_get_speed` function returns the current [GPS](#gps) speed in meters per second.
 
+The `wb_gps_get_speed_vector` function returns the current [GPS](#gps) speed vector (difference between current and previous position) in centimeters.
```
We rely on the metric system for units.
Therefore, the unit for the speed vector should be meters per second.

omichel commented on projects/samples/devices/controllers/gps/gps.c:
```diff
@@ -91,7 +92,12 @@ int main() {
-      default:
+      case 'V': {
+        const double *speed_vector_values = wb_gps_get_speed_vector(gps);
+        printf("Using the GPS device Speed Vector: %.3f %.3f %.3f\n", speed_vector_values[0], speed_vector_values[1], speed_vector_values[2]);
```
```suggestion
        printf("GPS speed vector: %.3f %.3f %.3f\n", speed_vector_values[0], speed_vector_values[1], speed_vector_values[2]);
```

llessieux replied:
Ok

omichel commented on src/webots/nodes/WbGps.cpp:
```diff
@@ -307,8 +307,9 @@ bool WbGps::refreshSensorIfNeeded() {
     mMeasuredPosition[2] = altitude;
   }
 
+  mSpeedVector = (t - mPreviousPosition);
```
`mSpeedVector` should be computed in meters per second.

llessieux replied:
Doh indeed.

llessieux commented:
A new gps_speed_vector test should probably be added but I am not familiar with what the default setup is doing.
So I am not sure what I should be expecting ;) I guess I can debug it and use the values as expected results but that is not really good ;)

omichel reviewed (changes requested):
For matlab, you will have to implement the method in [webots/lib/controller/matlab](https://github.com/cyberbotics/webots/tree/develop/lib/controller/matlab), it should be pretty similar to [wb_receiver_get_emitter_direction.m](https://github.com/cyberbotics/webots/blob/develop/lib/controller/matlab/wb_receiver_get_emitter_direction.m) which returns a vector of 3 floating point values.

And yes, a test should be added, probably in about [here](https://github.com/cyberbotics/webots/blob/develop/tests/api/controllers/gps_speed/gps_speed.c#L29).

llessieux commented:
Do you want me to add the tests in the same c file as the speed test or create a new one?

omichel commented:
> Do you want me to add the tests in the same c file as the speed test or create a new one?

I believe adding the test in the same file is fine.

llessieux commented:
About the matlab code, I have now added the function but I am confused by the .gitignore.
Should I add the function in there like the other ones or not?

llessieux commented:
Hi Olivier, I had a look at the mgenerate.py and it has the wb_gps_get_values commented.
I think that all functions returning pointers are commented. So I added mine commented too.
And added the new .m file for it, similar to the one you pointed me at.

omichel commented:
Yes, it would be good to add the ROS part indeed.

llessieux commented:
The ROS part, that's a separate repo correct? 
I will need a new PR for it if it is I guess. I'll start looking into that tomorrow.

ShuffleWire commented:
First, thank for taking in charge ! :) 
Second, according to https://github.com/cyberbotics/webots/issues/3660#issuecomment-916691844 it should be possible to get ride of the numerical derivative. I'm sure that it would avoid issue from this first order derivation (noise, inaccuracy and stuff...), especially if ODE provide those values already !
Granted, you made a nice job, and that is only a nice-to-have, which could be implemented later.

llessieux commented on tests/api/controllers/gps_speed/gps_speed.c:
```diff
@@ -8,25 +8,63 @@
   ts_assert_double_in_delta(speed, 5.0, 0.0001, "The speed value measured by the GPS should be 5.0 and not %lf after two steps",
                             speed);
 
+  r = wb_gps_get_speed_vector(gps);
+  for (i = 0; i < 3; i++)
+    // clang-format off
+    // clang-format 11.0.0 is not compatible with previous versions with respect to nested conditional operators
+    ts_assert_double_in_delta(r[i], e[i], 0.000001,
+                              "The %c value measured by the GPS should be %g and not %g after %d wb_robot_step(s)",
+                              i == 0 ? 'X' :
+                              i == 1 ? 'Y' :
+                                       'Z',
+                              e[i], r[i], 2);
+  // clang-format on
+
```
Ok, i'll update that.

llessieux commented on tests/api/controllers/gps_speed/gps_speed.c:
```diff
@@ -8,25 +8,63 @@
                          "The speed value measured by the GPS should be NaN and not %lf before the device is enabled", speed);
 
+  r = wb_gps_get_speed_vector(gps);
+  for (i = 0; i < 3; i++)
+    // clang-format off
+    // clang-format 11.0.0 is not compatible with previous versions with respect to nested conditional operators
+    ts_assert_double_equal(r[i], NAN, "The %c value measured by the GPS should be NaN and not %g before the device is enabled",
+                           i == 0 ? 'X' :
+                           i == 1 ? 'Y' :
+                                    'Z',
+                           r[i]);
+  // clang-format on
+
```
same here

omichel commented on tests/api/controllers/gps_speed/gps_speed.c:
```diff
@@ -54,16 +42,10 @@ int main(int argc, char **argv) {
-                                       'Z',
-                              e[i], r[i], 2);
-  // clang-format on
+  ts_assert_vec3_equal(
+    r[0], r[1], r[2], NAN, NAN, NAN,
+    "The speed vector value measured by the GPS should be [%f, %f, %f] not [%f, %f, %f] after %d wb_robot_step(s)", e[0], e[1],
+    e[2], r[0], r[1], r[2], 2);
```
I am not sure this is correct, you should have used `ts_assert_vec3_in_delta` against the `e` values instead of `NAN` here.

llessieux replied:
C&P :)
Fixing it.
By the way, I realized that some of the ROS bit are in this project too.
In the RosGps.cpp/hpp if I am not mistaken.
Do you want me to update that too.

omichel replied:
> Do you want me to update that too.

Yes, please.

llessieux replied:
Added. I will add the webots_ros part tomorrow or a bit later today.
In a branch of the develop branch I supposed ;) Same name.

llessieux replied:
Damn forgot to update the doc for the ROS part.
adding that now.

omichel replied:
No idea. I guess this is fine this way.

omichel replied:
My colleague @BenjaminHug8 will review the ROS part.

omichel commented on tests/api/controllers/gps_speed/gps_speed.c:
```diff
@@ -8,25 +8,45 @@
                             speed);
 
+  r = wb_gps_get_speed_vector(gps);
+  ts_assert_vec3_equal(
```
I meant:
```suggestion
  ts_assert_vec3_in_delta(
```

omichel replied:
Otherwise I am not sure the test will pass...

BenjaminHug8 commented:
Hi @llessieux,

With Olivier we decided to remove the topic `gps/speed` as the new one has the `fram_id` and contains more information.

Beside I just put the publishing of this new topic inside another function, as it was supposed to be.

omichel commented on projects/default/controllers/ros/RosGPS.cpp:
```diff
@@ -70,12 +65,9 @@ void RosGPS::publishValue(ros::Publisher publisher) {
+void RosGPS::publishAuxiliaryValue() {
   geometry_msgs::PointStamped value;
   value.header.stamp = ros::Time::now();
   value.header.frame_id = mRos->name() + '/' + RosDevice::fixedDeviceName();
```
Shouldn't we add a more specific `frame_id` here? Like:
```C++
value.header.frame_id = mRos->name() + '/' + RosDevice::fixedDeviceName() + "/speed_vector";
```

omichel commented on docs/reference/gps.md:
```diff
@@ -138,6 +141,7 @@ speed = wb_gps_get_speed(tag)
 | --- | --- | --- | --- |
 | `/<device_name>/values` | `topic` | [`sensor_msgs::NavSatFix`](http://docs.ros.org/api/sensor_msgs/html/msg/NavSatFix.html) for `WGS84` GPS coordinate system<br/><br/>or<br/><br/>[`geometry_msgs::PointStamped`](http://docs.ros.org/api/geometry_msgs/html/msg/PointStamped.html) for `local` GPS coordinate system | [`Header`](http://docs.ros.org/api/std_msgs/html/msg/Header.html) `header`<br/>[`sen ...
 | `/<device_name>/speed` | `topic` | webots_ros::Float64Stamped | [`Header`](http://docs.ros.org/api/std_msgs/html/msg/Header.html) `header`<br/>`float64 data` |
+| `/<device_name>/speed_vector` | `topic` | [`geometry_msgs::PointStamped`](http://docs.ros.org/api/geometry_msgs/html/msg/PointStamped.html) for `local` GPS speed vector in meters per second| [`Header`](http://docs.ros.org/api/std_msgs/html/msg/Header.html) `header`<br/>`float64 x` <br/>`float64 y`<br/>`float64 z`|
```
@BenjaminHug8: this should be probably be updated.

BenjaminHug8 replied:
Done !

omichel reviewed (approved):
Thank you.
